File: generators/fromJson.ts
import { DartType, Definition, Format } from "./types";
import { getDartTypeByFormat } from "./utils";

export const generateFromJsonMethod = (
  className: string,
  properties: Definition["properties"],
  requiredFields: string[]
) => {
  let code = `factory ${className}.fromJson(Map<String, dynamic> json) {\n`;
  code += `return ${className}(\n`;
  for (const propertyName in properties) {
    const isRequired = requiredFields.includes(propertyName);
    const dartType = getDartTypeByFormat(properties[propertyName].format);
    code += `${propertyName}: ${parseWrapper(
      dartType,
      properties[propertyName].format,
      `json['${propertyName}']`
    )},\n`;
  }
  code += `);\n`;
  code += `}\n\n`;
  return code;
};

export const parseWrapper = (
  dartType: DartType,
  format: Format,
  jsonValue: string
): string => {
  const isArray = format.includes("[]");
  if (isArray) {
    format = format.replace("[]", "") as Format;
  }
  let output = `${jsonValue} != null ?`;
  switch (dartType) {
    // ====================
    case "int":
      output += `${jsonValue} as int`;
      break;
    case "List<int>":
      output += `(${jsonValue} as List<dynamic>).map((v) => v as int).toList()`;
      break;
    // ====================
    case "BigInt":
      output += `BigInt.tryParse(${jsonValue}.toString()) as BigInt`;
      break;
    case "List<BigInt>":
      output += `(${jsonValue} as List<dynamic>).map((v) => BigInt.tryParse(v.toString()) as BigInt).toList()`;
      break;
    // ====================
    case "double":
      output += `double.tryParse(${jsonValue}.toString())`;
      break;
    case "List<double>":
      output += `(${jsonValue} as List<dynamic>).map((v) => double.tryParse(v.toString()) as double).toList()`;
      break;
    // ====================
    case "num":
      output += `num.tryParse(${jsonValue}.toString())`;
      break;
    case "List<num>":
      output += `(${jsonValue} as List<dynamic>).map((v) => num.tryParse(v.toString()) as num).toList()`;
      break;
    // ====================
    case "bool":
      output += `${jsonValue} as bool`;
      break;
    case "List<bool>":
      output += `(${jsonValue} as List<dynamic>).map((v) => v as bool).toList()`;
      break;
    // ====================
    case "String":
      output += `${jsonValue}.toString()`;
      break;
    case "List<String>":
      output += `(${jsonValue} as List<dynamic>).map((v) => v as String).toList()`;
      break;
    // ====================
    case "DateTime":
      if (format === "time without time zone") {
        output += `DateTime.tryParse("1970-01-01T\${${jsonValue}.toString()}") as DateTime`;
        break;
      }
      if (format === "time with time zone") {
        output += `DateTime.tryParse("1970-01-01T\${${jsonValue}.toString()}") as DateTime`;
        break;
      }
      output += `DateTime.tryParse(${jsonValue}.toString()) as DateTime`;
      break;
    case "List<DateTime>":
      if (format === "time without time zone") {
        output += `(${jsonValue} as List<dynamic>).map((v) => DateTime.tryParse("1970-01-01T\${v.toString()}") as DateTime).toList()`;
        break;
      }
      if (format === "time with time zone") {
        output += `(${jsonValue} as List<dynamic>).map((v) => DateTime.tryParse("1970-01-01T\${v.toString()}") as DateTime).toList()`;
        break;
      }
      output += `(${jsonValue} as List<dynamic>).map((v) => DateTime.tryParse(v.toString()) as DateTime).toList()`;
      break;
    // ====================
    case "Duration":
      output += `${jsonValue} as Duration`;
      break;
    case "List<Duration>":
      output += `(${jsonValue} as List<dynamic>).map((v) => v as Duration).toList()`;
      break;
    // ====================
    case "Map<String, dynamic>":
      output += `${jsonValue} as Map<String, dynamic>`;
      break;
    case "List<Map<String, dynamic>>":
      output += `(${jsonValue} as List<dynamic>).map((v) => v as Map<String, dynamic>).toList()`;
      break;
    // ====================
    default:
      console.log(dartType);
      output += `Something went wrong, please open an issue on this`;
      break;
  }
  // Add Null Default Value, This will enable the support for column selection
  output += ` : ${dartTypeDefaultNullValue(dartType)}`;
  return output;
};

const dartTypeDefaultNullValue = (dartType: DartType): string => {
  switch (dartType) {
    case "int":
      return "0";
    case "BigInt":
      return "BigInt.from(0)";
    case "double":
      return "0.0";
    case "num":
      return "0";
    case "bool":
      return "false";
    case "String":
      return "'0'";
    case "DateTime":
      return "DateTime.fromMillisecondsSinceEpoch(0)"; // Using Unix epoch as default
    case "Duration":
      return "Duration()"; // Assuming duration in milliseconds
    case "Map<String, dynamic>":
      return "{}";
    case "dynamic":
      return "null";
    case "List<int>":
    case "List<BigInt>":
    case "List<double>":
    case "List<num>":
    case "List<bool>":
    case "List<String>":
    case "List<DateTime>":
    case "List<Duration>":
    case "List<Map<String, dynamic>>":
    case "List<dynamic>":
      return "[]";
    default:
      return "null";
  }
};

File: cli/lib/generators/class/from_json.dart
import '../swagger/column.dart';
import '../swagger/table.dart';

String generateFromJsonMethod(
    String className, Table table, bool jsonbToDynamic) {
  final columns = table.columns;
  final StringBuffer code = StringBuffer();

  code.writeln('factory $className.fromJson(Map<String, dynamic> jsonn) {');
  code.writeln('return $className(');

  columns.forEach((columnName, columnDetails) {
    code.writeln(
        '$columnName: ${decodeFromJson(columnDetails, jsonbToDynamic)},');
  });

  code.writeln(');');
  code.writeln('}');
  code.writeln();

  return code.toString();
}

String decodeFromJson(Column columnDetails, bool jsonbToDynamic) {
  final postgresFormat = columnDetails.postgresFormat;
  String jsonValue = 'jsonn[\'${columnDetails.dbColName}\']';
  String jsonDecode = "";
  if (jsonbToDynamic &&
      (postgresFormat == 'jsonb' || postgresFormat == 'jsonb[]')) {
    if (postgresFormat == 'jsonb') {
      jsonDecode = '$jsonValue as dynamic';
    } else {
      jsonDecode = '($jsonValue as List<dynamic>)';
    }
  } else {
    switch (postgresFormat) {
      case 'smallint':
      case 'integer':
        jsonDecode = 'int.parse($jsonValue.toString())';

        break;
      case 'smallint[]':
      case 'integer[]':
        jsonDecode =
            '($jsonValue as List<dynamic>).map((v) => int.parse(v.toString())).toList()';
        break;

      case 'bigint':
        jsonDecode = 'BigInt.parse($jsonValue.toString())';
        break;
      case 'bigint[]':
        jsonDecode =
            '($jsonValue as List<dynamic>).map((v) => BigInt.parse(v.toString())).toList()';
        break;

      case 'real':
      case 'double precision':
        jsonDecode = 'double.parse($jsonValue.toString())';
        break;

      case 'real[]':
      case 'double precision[]':
        jsonDecode =
            '($jsonValue as List<dynamic>).map((v) => double.parse(v.toString())).toList()';
        break;

      case 'numeric':
        jsonDecode = 'num.parse($jsonValue.toString())';
        break;
      case 'numeric[]':
        jsonDecode =
            '($jsonValue as List<dynamic>).map((v) => num.parse(v.toString())).toList()';
        break;

      case 'json':
      case 'jsonb':
        jsonDecode = '$jsonValue as Map<String, dynamic>';
        break;
      case 'json[]':
      case 'jsonb[]':
        jsonDecode =
            '($jsonValue as List<dynamic>).map((v) => json.decode(v) as Map<String, dynamic>).toList()';
        break;
      case 'text':
      case 'character varying':
      case 'character':
      case 'uuid':
        jsonDecode = '$jsonValue.toString()';
        break;

      case 'text[]':
      case 'character varying[]':
      case 'character[]':
      case 'uuid[]':
        jsonDecode =
            '($jsonValue as List<dynamic>).map((v) => v.toString()).toList()';
        break;

      // case 'date':
      // case 'date[]':
      // case 'time without time zone':
      // case 'time without time zone[]':
      // case 'time with time zone':
      // case 'time with time zone[]':
      // case 'timestamp without time zone':
      // case 'timestamp without time zone[]':
      // case 'timestamp with time zone':
      // case 'timestamp with time zone[]':
      case 'date':
        jsonDecode = 'DateTime.parse($jsonValue.toString())';
        break;
      case 'date[]':
        jsonDecode =
            '($jsonValue as List<dynamic>).map((v) => DateTime.parse(v.toString())).toList()';
        break;
      case 'time without time zone':
      case 'time with time zone':
        jsonDecode = 'DateTime.parse("1970-01-01T\$${{jsonValue}}").toLocal()';
        break;
      case 'time without time zone[]':
      case 'time with time zone[]':
        jsonDecode =
            '($jsonValue as List<dynamic>).map((v) => DateTime.parse("1970-01-01T\$${"v"}").toLocal()).toList()';
        break;
      case 'timestamp without time zone':
      case 'timestamp with time zone':
        jsonDecode = 'DateTime.parse($jsonValue.toString())';
        break;
      case 'timestamp without time zone[]':
      case 'timestamp with time zone[]':
        jsonDecode =
            '($jsonValue as List<dynamic>).map((v) => DateTime.parse(v.toString())).toList()';
        break;

      case 'boolean':
        jsonDecode += '$jsonValue as bool';
        break;
      case 'boolean[]':
        jsonDecode +=
            '($jsonValue as List<dynamic>).map((v) => v as bool).toList()';
        break;

      case 'interval':
        jsonDecode = 'DurationFromString.fromString($jsonValue.toString())';
        break;

      case 'interval[]':
        jsonDecode =
            '($jsonValue as List<dynamic>).map((v) => DurationFromString.fromString(v.toString())).toList()';
        break;

      // These types will be prefixed with the schema PostGIS is installed in
      // (eg. extensions.geometry), and may optionally end with a specific
      // geometry type if specified for the column (eg. extensions.geometry(POINT),
      // extensions.geometry(LINESTRING)), so check that they contain name
      // anywhere in the string.
      case final String s when s.contains('geometry') && s.endsWith('[]'):
      case final String s when s.contains('geography') && s.endsWith('[]'):
        jsonDecode =
            '($jsonValue as List<String>).map((v) => GeometryBuilder.decodeHex($jsonValue.toString(), format: WKB.geometryExtended)).toList()';
        break;

      case final String s when s.contains('geometry'):
      case final String s when s.contains('geography'):
        jsonDecode =
            'GeometryBuilder.decodeHex($jsonValue.toString(), format: WKB.geometryExtended)';
        break;

      // NOT YET SUPPORTED TYPES ARE ENCODED TO STRINGS BY DEFAULT
      // NEED CONTRIBUTIONS TO SUPPORT THESE TYPES
      case 'bytea':
      case 'inet':
      case 'cidr':
      case 'macaddr':
      case 'macaddr8':
      case 'point':
      case 'line':
      case 'lseg':
      case 'box':
      case 'path':
      case 'polygon':
      case 'circle':
      case 'bit':
      case 'bit varying':
      case 'money':
      case 'tsquery':
      case 'tsvector':
      case 'xml':
      case 'pg_lsn':
      case 'pg_snapshot':
      case 'txid_snapshot':
        jsonDecode = '$jsonValue.toString()';
        break;
      case 'bytea[]':
      case 'inet[]':
      case 'cidr[]':
      case 'macaddr[]':
      case 'macaddr8[]':
      case 'point[]':
      case 'line[]':
      case 'lseg[]':
      case 'box[]':
      case 'path[]':
      case 'polygon[]':
      case 'circle[]':
      case 'bit[]':
      case 'bit varying[]':
      case 'money[]':
      case 'tsquery[]':
      case 'tsvector[]':
      case 'xml[]':
      case 'pg_lsn[]':
      case 'pg_snapshot[]':
      case 'txid_snapshot[]':
        jsonDecode =
            "($jsonValue as List<dynamic>).map((e) => e.toString()).toList()";
        break;
      default:
        // check if its a vector
        if (postgresFormat.contains("vector") ||
            postgresFormat.contains("VECTOR")) {
          jsonDecode = '$jsonValue.toString()';
          break;
        }

        // check if the type is an enum type
        if (columnDetails.isEnum) {
          String enumName = columnDetails.dartType;
          if (postgresFormat.contains("[]")) {
            return '$jsonValue != null ? $enumName.from($jsonValue.map((e) => ${enumName.replaceAll("List<", "").replaceFirst(">", "")}.values.byName(e.toString())).toList()) : []';
          } else {
            return '$jsonValue != null ? $enumName.values.byName($jsonValue.toString()) : $enumName.values.first';
          }
        }

        // If the type is not supported, we will fallback to string type
        if (postgresFormat.contains("[]")) {
          jsonDecode =
              '($jsonValue as List<dynamic>).map((e) => e.toString()).toList()';
        } else {
          jsonDecode = '$jsonValue.toString()';
        }
    }
  }

  if (columnDetails.isNullable) {
    return '$jsonValue != null ? $jsonDecode : null';
  }

  return '$jsonValue != null ? $jsonDecode : ${dartTypeDefaultNullValue(columnDetails)}';
}

String dartTypeDefaultNullValue(Column columnDetails) {
  switch (columnDetails.dartType) {
    case 'int':
      return '0';
    case 'BigInt':
      return 'BigInt.from(0)';
    case 'double':
      return '0.0';
    case 'num':
      return '0';
    case 'bool':
      return 'false';
    case 'String':
      return "''"; // Empty string as default
    case 'DateTime':
      return 'DateTime.fromMillisecondsSinceEpoch(0)';
    case 'Duration':
      return 'Duration()'; // Assuming duration in milliseconds
    case 'Map<String, dynamic>':
      return '<String, dynamic>{}';
    case 'List<int>':
      return '<int>[]';
    case 'List<BigInt>':
      return '<BigInt>[]';
    case 'List<double>':
      return '<double>[]';
    case 'List<num>':
      return '<num>[]';
    case 'List<bool>':
      return '<bool>[]';
    case 'List<String>':
      return '<String>[]';
    case 'List<DateTime>':
      return '<DateTime>[]';
    case 'List<Duration>':
      return '<Duration>[]';
    case 'List<Map<String, dynamic>>':
      return '<Map<String, dynamic>>[]';
    case 'Geometry':
      return 'Point(Position.view(const [double.nan, double.nan]))';
    case 'List<Geometry>':
      return '<Geometry>[]';
    default:
      return 'Unsupported Dart type: please open an issue';
  }
}
